### PR TITLE
[MAINTENANCE] Add `AddedDiagnostics` helper class to `is_added` flows

### PR DESCRIPTION
**Pull request by cdkini**
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/core/added_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    errors: list[ResourceNotAddedError]
+
+    @property
+    def is_added(self) -> bool:
+        return len(self.errors) == 0
+
+    @abstractmethod
+    def raise_for_error(self) -> None:
+        """
```
Comment (tyler-hoffman) [inline great_expectations/core/added_diagnostics.py:42]: So this will either raise or potentially return None?

Comment (cdkini) [reply great_expectations/core/added_diagnostics.py:42]: Yup akin to `requests.Response.raise_for_status`

Comment (cdkini) [reply great_expectations/core/added_diagnostics.py:42]: could do something like:
```
if not diagnostics.is_added:
    raise diagnostics.exception()
```
but the conditional check will be the same with every use

File: great_expectations/checkpoint/checkpoint.py [hunk 2]
```diff
@@ -158,13 +166,13 @@ def run(
...
-        if not added:
+        diagnostics = self.is_added()
+        if not diagnostics.is_added:
             # The checkpoint itself is not added but all children are - we can add it for the user
-            if len(errors) == 1 and isinstance(errors[0], CheckpointNotAddedError):
+            if not diagnostics.parent_added and diagnostics.children_added:
                 self._add_to_store()
             else:
-                raise CheckpointRelatedResourcesNotAddedError(errors=errors)
+                diagnostics.raise_for_error()
```
Comment (tyler-hoffman) [inline great_expectations/checkpoint/checkpoint.py:175]: @cdkini this line being in a conditional is the reason I was surprised below that it didn't always raise. Not saying anything needs to change. Just thinking aloud.

**Review by tyler-hoffman — APPROVED** [review]
I think this looks reasonable. One thing that we'll maybe have to think about in the future (or maybe you already thought about this) is that something can exist and have an id, but have been deleted. I think it's a weird state we don't have to solve for today.

**Pull request merged**